### PR TITLE
repart: support engines and providers

### PR DESCRIPTION
**Pull request by bluca**
As discussed, providers are too new and unstable, and do not yet work reliably, so engines have to be supported too.

### CODE REVIEW

File: src/shared/openssl-util.c
Comment (daandemeyer) [inline src/shared/openssl-util.c:1314]: Do we still need these? if I remember correctly `OSSL_STORE_open()` should load the provider itself no? (I know the whole provider interface with the file in /etc sucks but I think we can blissfully ignore that, we just try to use the provider interface as intended, and the fact it doesn't actually work isn't our problem, as long as we make sure the engine stuff does work as intended).

File: src/shared/openssl-util.c
Comment (daandemeyer) [inline src/shared/openssl-util.c]: `openssl_load_key()` since this falls back to engines as well?

Comment (daandemeyer) [reply src/shared/openssl-util.c]: Just name the return value `ret` since there's only one?

File: src/shared/openssl-util.c
Comment (daandemeyer) [inline src/shared/openssl-util.c]: Unconditionally falling back seems a bit icky, but I don't really see a better solution unless we can figure out in which cases it makes sense to fall back and in which cases it doesn't.

Maybe add an environment variable to force either usage of providers or engines and if not set, we do the current logic of trying provider first and engine as a fallback?

File: src/shared/openssl-util.c
Comment (daandemeyer) [reply src/shared/openssl-util.c]: Also we need an assert on the return argument here

File: src/shared/openssl-util.c
Comment (daandemeyer) [inline src/shared/openssl-util.c]: This seems weird, we generally don't free the existing value like this, so I'd drop this.

File: src/shared/openssl-util.c
Comment (daandemeyer) [inline src/shared/openssl-util.c]: Same here, I'd leave it to the caller to free any previous value

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks great, let me cc @poettering as well to take a look

File: src/shared/openssl-util.c
Comment (bluca) [reply src/shared/openssl-util.c:1314]: That would break the hermetic-usr specification - things should work in a sensible way by default with an empty /etc. If systemd failed to boot unless /etc/systemd/system.conf was present, it would be pretty lousy. There's no reason to fail here, it's trivial to make it work by default.

File: src/shared/openssl-util.c
Comment (bluca) [reply src/shared/openssl-util.c]: I can add an env var - when I have no provider, with my current version of openssl it doesn't fail to open the provider even if there isn't any, it only fails later when trying to load the key, so yeah it needs to be a generic fallback for it to work reliably.

File: src/shared/openssl-util.c
Comment (daandemeyer) [reply src/shared/openssl-util.c:1314]: Can we add a comment here explaining the reasoning for doing this? Also, why also load the "default" provider?

Ideally you'd also link to an issue in the openssl github repo that asks them to support having config files in both /usr and /etc

File: src/shared/openssl-util.c
Comment (daandemeyer) [reply src/shared/openssl-util.c]: We should also debug log here that we're falling back to engines

File: src/shared/openssl-util.c
Comment (bluca) [reply src/shared/openssl-util.c]: I've renamed to openssl_load_key_from_token - it's not just loading any key, it's from an exterrnal 'thing', so I think it needs some suffix to qualify it

The asserts are down the chain, as this function doesn't deref the pointer at all

File: src/shared/openssl-util.c
Comment (bluca) [reply src/shared/openssl-util.c:1314]: Loading the default as well seemed to be the suggestion from what I gathered when adding this. It doesn't hurt to do so.
I really do not have the energy to go fight with openssl, but I'll add a comment

File: src/shared/openssl-util.c
Comment (daandemeyer) [inline src/shared/openssl-util.c:1397]: Hmm, thinking more about it, I really dislike doing the fallback regardless of what error we get. Would it be OK for you if we by default used the providers and using the engine requires setting the environment variable? Passing through the environment variable to systemd-repart in mkosi is as simple as setting it in `Environment=`.

Comment (bluca) [reply src/shared/openssl-util.c:1397]: The providers do not work at all, so basically it means that other than the command line option, the env var must always be set for this to have any meaning and usefulness, otherwise it's basically dead code. Even when they eventually start to work at some indeterminate point in the future, there's never going to be a clear error, just "i/o error". At that point I would question why add support for providers in the first place, and would just do engines only instead, given they actually work

Comment (bluca) [reply src/shared/openssl-util.c:1397]: It's just some fallback for some external wonky API, why does it even matter?

**Comment by poettering** [comment]
> As discussed, providers are too new and unstable, and do not yet work reliably, so engines have to be supported too.

Which discussion are you referring to, precisely?

**Comment by bluca** [comment]
> > As discussed, providers are too new and unstable, and do not yet work reliably, so engines have to be supported too.
> 
> Which discussion are you referring to, precisely?

Between myself and Daan at FOSDEM

File: src/shared/openssl-util.c
Comment (poettering) [inline src/shared/openssl-util.c]: move to openssl-util.h to the other cleanup macros?

Comment (bluca) [reply src/shared/openssl-util.c]: That makes it exported, no? Stuff used only in a module is defined in the module usually I think?

File: src/shared/openssl-util.c
Comment (poettering) [inline src/shared/openssl-util.c]: please suffix with `;`, in order not to confuse editors

File: src/shared/openssl-util.c
Comment (poettering) [inline src/shared/openssl-util.c]: suffix with `;`

File: src/shared/openssl-util.c
Comment (bluca) [reply src/shared/openssl-util.c]: gcc 12 didn't complain here - but sure

File: docs/ENVIRONMENT.md
Comment (poettering) [inline docs/ENVIRONMENT.md:135]: maybe say explicitly that "provider API" is the way to go and the future, but that reality bites and hence "engine API" might have to be used instead. I.e. I think there#s value in explaining the situation briefly, since it's not obvious for the casual onlooker

File: src/shared/openssl-util.c
Comment (bluca) [reply src/shared/openssl-util.c]: but it doesn't matter, I'll just move it

File: src/partition/repart.c
Comment (poettering) [inline src/partition/repart.c]: urks, i really don't like ambiguities like this whenever we can. We so far managed to do without any ambiguity like this.

I'd prefer if we'd expose this under some non-ambiguous scheme, maybe pair `--private-key=` with a new setting `--private-key-url=` or so? I mean, at least in the pkcs11 case you specify something you'd call an "url" there, don't you? is that generally the case for other relevant modules too? i.e. tpm engine? i have no idea how engines/provides work in practice? just asking?

**Comment by poettering** [comment]
(btw, that main track tpm dude at fosdem was pretty annoyed about the provider vs. engine situation too, and didn#t make the switch yet for his engine.)

File: src/partition/repart.c
Comment (bluca) [reply src/partition/repart.c]: I had a separate parameter for the provider that gated this logic, but @DaanDeMeyer asked me to remove it in the previous PR.

In general, for example in the openssl binary itself, you use the same private-key field that you would use to pass the plaintext key file, so this is consistent with the common approach that I can see.

With openssl you'd do something like: `openssl smime -sign -inkey 'pkcs11:....' -engine pkcs11`

Comment (bluca) [reply src/partition/repart.c]: I don't mind at all whether there's a new command line parameter or not, so just pick one scheme you both agree with and I'll implement it

Comment (poettering) [reply src/partition/repart.c]: the openssl cmdline leaves no ambiguity, because you pick the mode explicitly via -engine...

the problem i really have is with the ambiguity. i.e. "pkcs11:foo" could be a filename, could be a url, you can't know. hence i think `--private-key-url=` would be nicest I think, it avoids the ambiguity.

Comment (bluca) [reply src/partition/repart.c]: I've moved it to --private-key-uri

File: man/systemd-repart.xml
Comment (poettering) [inline man/systemd-repart.xml]: maybe say

```
Takes a URI-like string referring to a private key, that will be passed to OpenSSL's "engine" or "provider" logic.
```

File: src/partition/repart.c
Comment (poettering) [inline src/partition/repart.c]: doesn't matter, but here i'd actually use `else if...`

File: src/partition/repart.c
Comment (daandemeyer) [reply src/partition/repart.c]: As long as `--private-key-uri` can do everything and you don't have to use either --private-key or --private-key-uri depending on whether you want a file or not I'm OK with it.

Comment (bluca) [reply src/partition/repart.c]: Lennart wants a separate parameter for files, and a separate parameters for URIs. You want one parameter to accept both. I cannot do both things at the same time.

**Comment by DaanDeMeyer** [comment]
Still not a huge fan but I'm not going to block this either. I think instead of testing this with softhsm we'd be better off using something hardware emulated in qemu. Either using a smartcard (https://www.qemu.org/docs/master/system/devices/ccid.html) + OpenSC + PKCS11, tpm + PKCS11 or tpm + tpm2-openssl. But I have a feeling implementing such a test is going to be super messy, so I don't think we should block merging this on that.

File: src/partition/repart.c
Comment (daandemeyer) [reply src/partition/repart.c]: @bluca I mean that we can have both parameters, but `--private-key-uri` should also work if `file:/a/b/c` is specified, which I think will work by default, but it'd be nice if you checked that worked

Comment (bluca) [reply src/partition/repart.c]: It doesn't work, no idea how to make that work - or do you mean to change this to manually check if the uri starts with file: and if so redirect to the old file-based pre-existing parameter?

**Comment by poettering** [comment]
> Still not a huge fan

not a fan of what part precisely?

**Comment by poettering** [comment]
> I think instead of testing this with softhsm we'd be better off using something hardware emulated in qemu. Either using a smartcard (https://www.qemu.org/docs/master/system/devices/ccid.html) + OpenSC + PKCS11, tpm + PKCS11 or tpm + tpm2-openssl. But I have a feeling implementing such a test is going to be super messy, so I don't think we should block merging this on that.

So I know this trick, it works wonders: merge without CI test and wait that a certain @mrc0mmand might eventually find interest in it, and adds a really good test for you. has worked for me a couple of times. ;-)

File: src/partition/repart.c
Comment (poettering) [reply src/partition/repart.c]: i mean, we already released this stuff right, requiring file: would be quite a compat break...

I think in future we should maybe follow the rule for similar cases that fs path should be prefixed with "/" or "./", and that strings prefixed with neither are always something else. or in other words: if you explicitly specify a relative path or an absolute path it's a path, otherwise something else.

but here it's kinda too late...

Comment (poettering) [reply src/partition/repart.c]: oh, i get what @DaanDeMeyer suggests. and I think that would be fine, i.e. manually parse the `file:/` prefix and do the right thing, when passed in the *new* switch `--private-key-uri=`

That said, parsing `file:/` is kinda nasty. Just look at the ambiguities already listed on the wikipedia page:

https://en.wikipedia.org/wiki/File_URI_scheme

Comment (poettering) [reply src/partition/repart.c]: https://datatracker.ietf.org/doc/rfc8089/

**Comment by mrc0mmand** [comment]
> > I think instead of testing this with softhsm we'd be better off using something hardware emulated in qemu. Either using a smartcard (https://www.qemu.org/docs/master/system/devices/ccid.html) + OpenSC + PKCS11, tpm + PKCS11 or tpm + tpm2-openssl. But I have a feeling implementing such a test is going to be super messy, so I don't think we should block merging this on that.
> 
> So I know this trick, it works wonders: merge without CI test and wait that a certain @mrc0mmand might eventually find interest in it, and adds a really good test for you. has worked for me a couple of times. ;-)

I actually have a branch somewhere, that integrates [CanoKey](https://www.qemu.org/docs/master/system/devices/canokey.html) into our test suite, but it's annoying (even though it works pretty well), since you have to rebuild QEMU to make it work.

File: src/partition/repart.c
Comment (bluca) [reply src/partition/repart.c]: The previous parameter is there and still works, I really don't want to get into parsing that stuff by hand

Comment (poettering) [reply src/partition/repart.c]: i sympathize. i think the file:// stuff could be added later, if we want to. no need to complicate things with that now.

**Pull request merged**
